### ISSUE TITLE
HikariPool should consider the Connection/Login Timeout Parameter, if present in JDBC URL

### ISSUE THREAD

**Issue by LokeshAlamuri** [brettwooldridge/HikariCP#2092]
A strange issue is possible with the IBM Db2 JDBC driver (IBM JCC JDBC 4 Driver) while trying to fetch the connection from the IBM DB2 server. HikariPool's default 'connectionTimeout'  value is 30 seconds. If this value is not overridden by the configuration file or through API, we are setting aggressively the actual java.sql.DataSource 'LoginTimeout' to 30 seconds. During this process we are not checking the 'loginTimeout' ( or equivalent properties provided by other DB vendors) property present in the JDBC URL or in the Datasource properties.

**Please find the IBM documentation about 'loginTimeout' parameter.**

https://www.ibm.com/docs/en/db2/9.7?topic=pdsdjs-common-data-server-driver-jdbc-sqlj-properties-all-database-products

The maximum time in seconds to wait for a connection to a data source. After the number of seconds that are specified by loginTimeout have elapsed, the driver closes the connection to the data source. The data type of this property is int. The default is 0. A value of 0 means that the timeout value is the default system timeout value.

Sample code:

        HikariConfig hikariConfig = new HikariConfig();
        hikariConfig.setJdbcUrl("jdbc:db2://localhost:5555/STLEC1:loginTimeout=10000;");
        hikariConfig.setUsername("test");
        hikariConfig.setPassword("pass");
        hikariConfig.setDriverClassName("com.ibm.db2.jcc.DB2Driver");

        HikariDataSource hikariDataSource = new HikariDataSource(hikariConfig);
        hikariDataSource.getConnection();

HikariCP Version: 4.0.3

		
Since this issue occurs exactly during connection fetching it is difficult to reproduce. The typical exception stack trace looks as below:

at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:138) 
	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:364)
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:206)
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:476)
	at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:561) 
	at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:115)
	at com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:81)
	at com.example.demo.HikariDataSourceTest.main(HikariDataSourceTest.java:24)

Caused by: java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:171)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	
	

The biggest problem here is, 

**When we try to fetch the connection for the first time 'loginTimeout' property specified in the JDBC URL is applied. This is because, we are not applying the pool connectionTimeout during 'com.zaxxer.hikari.pool.HikariPool.checkFailFast()' API execution. As per the defined DB server SLA we will be able to fetch the connection and it is added to the pool. When we are trying the fetch the second connection, Hikari will apply the pool Timeout. So, we see an awkward exception in the middle of the application lifecycle**.

I am not saying that we could configure this value, but many applications still live on the default values of popular frameworks. Keeping in view of our large community users, I feel we should resolve this ambiguity and make it much easier to use.

Apart from that default 30 seconds connection timeout would be too low for many heavily loaded DB servers. It would be a nice idea to increase this Timeout to a reasonable value like 5 minutes.

In summary, I would request you to consider any one or all of the following changes.

1) Consider 'loginTimeout' property in DB2 JDBC URL before overriding the actual datasource 'LoginTimeout'.
2) During 'com.zaxxer.hikari.pool.HikariPool.checkFailFast()' API fetch the connection with the configured connectionTimeout or fetch the connection without Timeout but do not add to the pool.
3) Consider increasing the default 30 seconds connection timeout value.

**Comment by LokeshAlamuri** [comment]
Could someone please take a look at this issue? Please let me know if there are any questions regarding the issue I have reported.

**Comment by LokeshAlamuri** [comment]
@brettwooldridge Could you please take a look at the bug I have reported? Please give your suggestions.

**Comment by LokeshAlamuri** [comment]
@pjmatt @lfbayer
Could you please take a look at the bug I have reported? Please give your suggestions.

**Comment by lfbayer** [comment]
HikariCP doesn't care what the URL looks like and the names of the properties is not part of the JDBC spec so HikariCP shouldn't know the specific properties that each driver might implement.

This doesn't sound like a HikariCP issue. Does setting the connectionTimeout on the HikariCP config mitigate the issue?

**Comment by LokeshAlamuri** [comment]
> HikariCP doesn't care what the URL looks like and the names of the properties is not part of the JDBC spec so HikariCP shouldn't know the specific properties that each driver might implement.
> 
> This doesn't sound like a HikariCP issue. Does setting the connectionTimeout on the HikariCP config mitigate the issue?

We are overriding the LoginTimeout of DataSource with 30 seconds default value.  HikariCP is wrapping around Datasource and providing pooled connections. If we are overriding its default value i.e actual Datasource, we should consider the properties in JDBC URL. Because, user might actually expect that the timeout would work as per the parameter specified in the JDBC URL, when we take HikariCP with default configuration.

If we are not planning to care about jdbc parameters, during 'com.zaxxer.hikari.pool.HikariPool.checkFailFast()' API we are fetching the connection without connectionTimeout and we are adding it to the pool. But, for the second connection it may throw timeout error since we can not retrieve the connection as per the DB SLA. Here, it is not consistent.

I shall provide you more details if required.

**Comment by LokeshAlamuri** [comment]
I think here we should consider JDBC URL parameters, since we are setting the LoginTImeout on java.sql.Datasource. As a wrapper around the actual DataSource it is our responsibility to consider the parameters what we are overriding. Please suggest, if we take it as a Jira and work on it.

**Comment by LokeshAlamuri** [comment]
> HikariCP doesn't care what the URL looks like and the names of the properties is not part of the JDBC spec so HikariCP shouldn't know the specific properties that each driver might implement.
> 
> This doesn't sound like a HikariCP issue. Does setting the connectionTimeout on the HikariCP config mitigate the issue?

One possible solution could be check from the JDBC URL,  if the DB is IBM DB2 and then take the loginTimeout parameter from the JDBC URL, Set it as 'LoginTimeout' for  java.sql.DataSource and connectionTimeout (poolTimeout)

**Comment by LokeshAlamuri** [comment]
This is the API where we are setting the LoginTimeout for Datasource with default connectionTimeout

/**
    * Set the loginTimeout on the specified DataSource.
    *
    * @param dataSource the DataSource
    */
   private void setLoginTimeout(final DataSource dataSource)
   {
      if (connectionTimeout != Integer.MAX_VALUE) {
         try {
            dataSource.setLoginTimeout(Math.max(1, (int) MILLISECONDS.toSeconds(500L + connectionTimeout)));
         }
         catch (Exception e) {
            logger.info("{} - Failed to set login timeout for data source. ({})", poolName, e.getMessage());
         }
      }
   }

**Comment by LokeshAlamuri** [comment]
> HikariCP doesn't care what the URL looks like and the names of the properties is not part of the JDBC spec so HikariCP shouldn't know the specific properties that each driver might implement.
> 
> This doesn't sound like a HikariCP issue. Does setting the connectionTimeout on the HikariCP config mitigate the issue?

No the problem would not be resolved. 'loginTimeout' parameter specified in JDBC URL will always takes precedence. So, we need to explicitly set the 'connectionTimeout' with the value on the HikariCP config.  The problem is when we are working with default values and having 'loginTimeout' parameter specified in JDBC UR. At that point if we want to make things work properly (without explicitly adding 'connectionTimeout' in config file), we have to read the 'loginTimeout' parameter from the DB2 JDBC URL and set it as connectionTimeout.
Internally, we are creating 'DriverDataSource' and setting the connectionTimeout value using the API
DriverManager.setLoginTimeout(seconds);

Many users may not be aware of this, it creates inconsistency as mentioned in my earlier comments. It also creates exceptions in the middle of application lifecycle.

**Comment by LokeshAlamuri** [comment]
@pjmatt @lfbayer
As mentioned in my previous comments, we are overriding the 'LoginTimeout' using the 

**'DriverManager.setLoginTimeout(seconds);'**


Since we have stepped in to this complexity, I feel we should consider JDBC URL parameter even though this is not part of the spec. We need not bother all the parameters but we should consider the equivalent parameters that we are overriding. Also, this is not the only issue where we consider the specifics of DB vendor. If we see the following API, we can see that we have considered a bug in MySQL and provided a patch on our side.

 **private void createNetworkTimeoutExecutor(final DataSource dataSource, final String dsClassName, final String jdbcUrl)**


If we don't want to consider the JDBC parameters, at least we should ensure that those specific parameters do not leave the HikariCP in buggy state. As mentioned in my previous comments, we are retrieving the first connection without 'connectionTimeout' and here 'loginTimeout' is enforced. But, for the second connection we are timing out the pool based on 'connectionTimeout'. 
If additional details are required, I shall provide you. **Please consider accepting this bug and let us have a discussion. Based on the suggestions from the contributors I shall close this finding.**

Please give your valuable suggestions.

**Comment by LokeshAlamuri** [comment]
@brettwooldridge @pjmatt @lfbayer 
Gentle reminder. Could you please give your suggestions. If you suggest not fixing this issue, please consider updating our wiki and advising the user community about the possible problem, if we use connectionTimeout parameters in JDBC URL.
